### PR TITLE
Fix attention parity for GPT-2

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**: 

Update attention operator to be in parity with huggingface GPT-2 & PyTorch.
(1) Use huggingface logic for unidirectional (causal) mask. Previous implementation add -10000.0, and huggingface will replace the value with -10000.0 for causal mask. This impacts the value of words masked by attention mask, and has slight impact on non masked words.
(2) Use persistent softmax in cuda to have same result as PyTorch. This seems to be major cause of fp16 difference.

After this change, CUDA result is 100% same. However CPU result still have slight difference from GEMM output.

For testing, add an environment variable (ORT_TRANSFORMER_OPTIONS=1 or 2) to enable/disable persistent softmax in attention.

Test Results:
* optimized=True, batch_size=2, hidden_size=4096, num_attention_heads=32, sequence_length=3, past_sequence_length=5, float16=False, padding_length=2, device=cuda
    * Before: Max_diff=1.7881393432617188e-07; Diff_count=100/100
    * After: Max_diff=0.0; Diff_count=0/100
* optimized=True, batch_size=2, hidden_size=4096, num_attention_heads=32, sequence_length=3, past_sequence_length=5, float16=True, padding_length=2, device=cuda
    * Before: Max_diff=0.00048828125; Diff_count=100/100
    * After: Max_diff=0.0; Diff_count=0/100
* optimized=True, batch_size=64, hidden_size=768, num_attention_heads=12, sequence_length=3, past_sequence_length=5, float16=False, padding_length=2, device=cpu
    * Before: Max_diff=1.862645149230957e-07; Diff_count=10/10
    * After: Max_diff=1.7881393432617188e-07; Diff_count=10/10

Average latency (ms) on GPT-2 (12 layers 768 hidden) with v100 GPU or CPU only:
| batch_size | sequence length | past sequence length | precision | GPU Latency (After) | GPU Latency (Before) | CPU Latency (After) | CPU Latency (Before) |
| ------- | -------  | -------  | -------  | -------  | -------  | -------  | -------  |
| 1 | 1 | 8 | FP32 | 2.57 | 2.48 | 10.35 | 10.39|
| 1 | 1 | 16 | FP32 | 2.54 | 2.52 | 10.59 | 10.09| 
| 1 | 1 | 32 | FP32 | 2.53 | 2.52 |  10.71 | 11.04|
| 1 | 1 | 64 | FP32 | 2.54 | 2.55 |  11.70 | 11.67|
| 1 | 1 | 128 | FP32 | 2.55 | 2.53 | 13.30 | 12.74|
| 1 | 1 | 8 | FP16 | 1.82 | 1.80 | NA | NA|
| 1 | 1 | 16 | FP16 | 1.81 | 1.81 | NA | NA|
| 1 | 1 | 32 | FP16 | 1.82 | 1.77 | NA | NA|
| 1 | 1 | 64 | FP16 | 1.82 | 1.82 | NA | NA|
| 1 | 1 | 128 | FP16 | 1.81 | 1.78 | NA | NA|

Using persistent softmax usually is slower (around 1%-3%). We use it by default for GPT-2, but not BERT.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

Related issues: https://github.com/microsoft/onnxruntime/issues/6549

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h [hunk 1]
```diff
@@ -179,30 +182,36 @@ __device__ inline void SoftmaxWithRawMaskSmall(const int all_sequence_length,
...
+    const int batch_index = blockIdx.y;
     if (mask_dimension == 2) {
       mask_offset = batch_index * all_sequence_length + threadIdx.x;
     } else if (mask_dimension == 3) {
       mask_offset = (batch_index * sequence_length + sequence_index) * all_sequence_length + threadIdx.x;
-    } else if (mask_dimension == 4){
-      // Megatron code:
-      // ltor_mask = ltor_mask[..., (attention_scores.size(3)-hidden_states.size(1)):attention_scores.size(3), :attention_scores.size(3)]
+    } else if (mask_dimension == 4) {
       mask_offset = (batch_index * max_sequence_length + all_sequence_length - sequence_length + sequence_index) * max_sequence_length + threadIdx.x;
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:202]: Can we keep the comment here since it's a bit difficult to infer the original logic when looking at the cuda code for mask_dim==4

**Review by wangyems — APPROVED** [review]

File: onnxruntime/contrib_ops/cuda/bert/attention_softmax.h (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/bert/attention_softmax.h:202]: @wangyems, I did not find the code in https://github.com/NVIDIA/Megatron-LM/tree/main/megatron.
